agent/cache/ratelimit.go

```diff
@@ -0,0 +1,113 @@
+package cache
+
+import (
+	"errors"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+)
+
+const (
+	// RateLimiterHasBeenStopped is error returned when rate-limiter has been stopped
+	RateLimiterHasBeenStopped = "RateLimiter is Stopped"
+)
+
+// RateLimiter is the interface to have a Rate-Limiter object
+type RateLimiter interface {
+	Take() error
+	Stop() error
+}
+
+// fake RateLimiter when ticker would cost too much (too small delay)
+type rateLimiterFake struct {
+	stopped bool
+}
+
+// RateLimiterTicker let the user Control request at a given rate
+type rateLimiterTicker struct {
+	// The ticker
+	ticker *time.Ticker
+	// Stop channel
+	stop  chan struct{}
+	burst chan struct{}
+}
+
+// NewRateLimiter builds a new Rate Limiter
+func NewRateLimiter(spec *lib.RateLimitSpec) RateLimiter {
+	if spec == nil || spec.Period < 1 {
+		return &rateLimiterFake{}
+	}
+	ticker := time.NewTicker(spec.Period)
+	stop := make(chan struct{}, 1)
+	burst := make(chan struct{}, spec.BurstSize)
+	rateLimiter := &rateLimiterTicker{
+		ticker,
+		stop,
+		burst,
+	}
+	rateLimiter.start()
+	return rateLimiter
+}
+
+// Take wait until rate-limiter allows you to continue
+func (r *rateLimiterTicker) Take() error {
+	ticker := r.ticker
+	if ticker == nil {
+		return errors.New(RateLimiterHasBeenStopped)
+	}
+	for {
+		select {
+		case <-r.burst:
+			return nil
+		}
+	}
+}
+
+// start is internal only as the ticker
+// is started by default
+func (r *rateLimiterTicker) start() {
+	go func() {
+		for {
+			select {
+			case <-r.stop:
```

Comment:
I haven't looked over this implementation in detail yet, but I was surprised to see it using a ticker and a goroutine.

I think it might be possible to track counts using buckets of time, and then remove old buckets after the time has passed. I think that is the algorithm used in https://github.com/juju/ratelimit. I haven't thought too much about the trade-offs here. What do you see as the advantage to this goroutine+ticker approach?

I'm not sure if it would make sense to use https://github.com/juju/ratelimit, or maybe we could implement something smaller based on the same algorithm.

Reply:
Or how about https://godoc.org/golang.org/x/time/rate#Limiter ? Would that work? I think we already have `x/time` as a dependency, so it wouldn't require adding a new one.

Reply:
I did not know about x/time/rate and I asked @banks whether I should be using a 3rd party or implement my own, he told me to implement my own to avoid having too many dependencies, so I did :-)

I can switch to x/time/rate if you prefer, I am ok with that as it has all the required stuff in it

Reply:
Perhaps we missunderstood each other Pierre but I certainly meant to say that we should just use `rate.Limit` just like we do everywhere else in the code 😄 . Apologies if that wasn't clear.

E.g. 
https://github.com/hashicorp/consul/blob/51c3a605ad4fc678a08901c107483203319f5949/agent/consul/config.go#L20
https://github.com/hashicorp/consul/blob/51c3a605ad4fc678a08901c107483203319f5949/agent/consul/config.go#L433

Reply:
ok, I am revamping this then.

Reply:
Done in https://github.com/hashicorp/consul/pull/8226/commits/bcb2fd7bff756c1535cea41d513731079c04c5a7